mixnet/node.py

```diff
@@ -1,107 +1,159 @@
 from __future__ import annotations
 
 import asyncio
-from contextlib import suppress
-from typing import Self, Tuple, TypeAlias
+from enum import Enum
+from typing import TypeAlias
 
-from cryptography.hazmat.primitives.asymmetric.x25519 import (
-    X25519PrivateKey,
-)
 from pysphinx.sphinx import (
     Payload,
     ProcessedFinalHopPacket,
     ProcessedForwardHopPacket,
     SphinxPacket,
-    UnknownHeaderTypeError,
 )
 
-from mixnet.config import MixNodeConfig, NodeAddress
-from mixnet.poisson import poisson_interval_sec
+from mixnet.config import GlobalConfig, NodeConfig
+from mixnet.connection import DuplexConnection, MixSimplexConnection
+from mixnet.gossip import GossipChannel
+from mixnet.packet import Fragment, MessageFlag, MessageReconstructor, PacketBuilder
 
-PacketQueue: TypeAlias = "asyncio.Queue[Tuple[NodeAddress, SphinxPacket]]"
-PacketPayloadQueue: TypeAlias = (
-    "asyncio.Queue[Tuple[NodeAddress, SphinxPacket | Payload]]"
-)
+BroadcastChannel: TypeAlias = asyncio.Queue[bytes]
 
 
-class MixNode:
+class Node:
     """
-    A class handling incoming packets with delays
-
-    This class is defined separated with the MixNode class,
-    in order to define the MixNode as a simple dataclass for clarity.
+    This represents any node in the network, which:
+    - generates/gossips mix messages (Sphinx packets)
+    - performs cryptographic mix (unwrapping Sphinx packets)
+    - generates noise
     """
 
-    config: MixNodeConfig
-    inbound_socket: PacketQueue
-    outbound_socket: PacketPayloadQueue
-    task: asyncio.Task  # A reference just to prevent task from being garbage collected
-
-    @classmethod
-    async def new(
-        cls,
-        config: MixNodeConfig,
-    ) -> Self:
-        self = cls()
+    config: NodeConfig
+    global_config: GlobalConfig
+    mixgossip_channel: GossipChannel
+    reconstructor: MessageReconstructor
+    broadcast_channel: BroadcastChannel
+    packet_size: int
```

Comment:
Shouldn't we extract the `packet_size` as a global parameter? Similar to the GTR.

Reply:
I thought about it. But, I realized that it's hard to estimate the actual packet size when configuring nodes, because the actual packet size is determined by the Sphinx packet encoding protocol, which depends on the max length of mix path (in the global config). Of course, we can provide a equation for users to calculate the actual packet size based on the max path length. But, I think it doesn't provide better user experience. Instead, I decided to calculate the actual packet size automatically when the node is initialized, so that it can be used when generating noise packet (in the same size as the real packet, which is a Sphinx).

Reply:
So maybe we could define global `max_message_size` which then will be used to calculate the `sphinx_packet_size` (I'm using this instead of `packet_size` as it is more precise to me)? I'm imagining that it should be something like `sphinx_packet_size = sphinx_header_size(max_mix_hops) + max_message_size`.

Reply:
The calculation of `sphinx_packet_size` is little bit more complex. So yeah, it is good to have the global parameter `max_message_size`, but I’d suggest to calculate the `sphinx_packet_size` internally by creating a sample sphinx packet instead of calculating an equation that can be probably changed by the Sphinx library. Of course, we’ll maintain the Sphinx library ourselves. We can later add a function in the library to calculate the size without creating a sample

Reply:
Yes, you are right, it's better not to design a function to calculate the `sphinx_packet_size` by hand and getting that value from sample sphinx packet.

Reply:
c360a4a I just added a new global parameter `max_message_size`. If a node tries to send a message longer than that, a `ValueError` will occur. All messages shorter than that will be padded by the sphinx encoding to make all sphinx packets have the same size.

I just found that the current Sphinx library (that I implemented) have a hard-coded max message size (that is defined by Loopix Sphinx). I'll make it flexible, so that our `max_message_size` parameter is used for that.